src/eval.c

```diff
@@ -3665,6 +3674,367 @@ get_lit_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+ * Or return NULL if evaluating is FAIL.
+ */
+    static char_u*
+eval_template_current(garray_T *current, int evaluate)
+{
+    typval_T	result;
+    char_u	*to_eval_current = (char_u *) current->ga_data;
+
+    if (!eval1(&to_eval_current, &result, evaluate))
+    {
+	ga_clear(current);
+	return NULL;
+    }
+    ga_clear(current);
+
+    return result.vval.v_string;
+}
+
+/*
+ * Consumes current to evaluate,
+ * and joins its result into result.
+ *
+ * Please also see eval_template_current().
+ */
+    static int
+eval_template_current_into_result(
+	garray_T *current,
+	garray_T *result,
+	char_u quote,
+	int evaluate)
+{
+    char_u  *current_result;
+    int	    is_eval_success;
+
+    current_result = eval_template_current(current, evaluate);
+    if (current_result == NULL)
+    {
+	return FAIL;
+    }
+    ga_concat_free(result, current_result);
+
+    return OK;
+}
+
+/*
+ * Simular to ga_concat(),
+ * but to_free will be vim_free() after ga_concat().
+ */
+    static void
+ga_concat_free(garray_T *ga, char_u *to_free)
+{
+    ga_concat(ga, to_free);
+    vim_free(to_free);
+}
+
+/*
+ * Allocate a variable for $"${expr}" and $'${expr}' constant.
+ * Return OK or FAIL.
+ */
+    static int
+get_template_string_tv(char_u **arg, typval_T *rettv, int evaluate)
+{
+    garray_T	result;  // a string that is not a template string
+    garray_T	current;
+    char_u	*to_eval_current;
+    typval_T	current_result;
+    char_u	*arg_orig = *arg;
+    char_u	quote = *(*arg + 1);
+    int		is_literal_string = (quote == '\'');
+    int		is_skipping_needed;
+
+    *arg += 2;  // $'
+
+    ga_init2(&result, 1, 80);
+
+    ga_init2(&current, 1, 80);
+    ga_append(&current, quote);
+
+    // Continue while it is not NULL and it is not a closing quote.
+    for (; (**arg != NUL) &&
+	    !is_closing_quote(*arg, is_literal_string, &is_skipping_needed);
+	    ++*arg)
+    {
+	if (is_skipping_needed)
+	{
+	    ga_append(&current, (int) **arg);
+	    ++*arg;
+	    ga_append(&current, (int) **arg);
+	    continue;
+	}
+	else if (**arg == '$' && *(*arg + 1) == '{')
+	{
+	    int is_eval_expr_success;
+	    int is_eval_current_success;
+
+	    ga_append(&current, quote);  // close quotation
+	    is_eval_current_success = eval_template_current_into_result(
+		    &current,
+		    &result,
+		    quote,
+		    evaluate);
+	    ga_append(&current, quote);  // open quotation
+
+	    if (!is_eval_current_success)
+	    {
+		ga_clear(&current);
+		ga_clear(&result);
+		return FAIL;
+	    }
+
+	    is_eval_expr_success =
+		eval_next_template_expr_into_result(
+			&result,
+			arg,
+			is_literal_string,
+			evaluate);
+	    if (!is_eval_expr_success)
+	    {
+		ga_clear(&current);
+		ga_clear(&result);
+		return FAIL;
+	    }
+
+	    continue;
+	}
+	ga_append(&current, (int) **arg);
+    }
+
+    /*
+     * Return FAIL if this template string is missing closing quote.
+     */
+    if (**arg == NUL)
+    {
+	ga_clear(&result);
```

Comment:
Isn't `ga_clear(&current);` needed here?

Reply:
Just now I 'm really stuck by this memory leak.
Really thanks!